cpg-core/src/main/kotlin/de/fraunhofer/aisec/cpg/passes/ImportResolver.kt

```diff
@@ -36,82 +39,85 @@ import de.fraunhofer.aisec.cpg.graph.scopes.NameScope
- * unit (usually in the form of a [NamespaceDeclaration]). The idea is to provide a sorted list of
- * TUs in which to resolve symbols and imports ideally. This is stored in [sortedTranslationUnits]
- * and is automatically computed the fist time someone accesses the property.
+ * This class holds the information about import dependencies between nodes that represent some kind
+ * of "module" (usually either a [TranslationUnitDeclaration] or a [Component]). The dependency is
+ * based on which module imports symbols of another module (usually in the form of a
+ * [NamespaceDeclaration]). The idea is to provide a sorted list of modules in which to resolve
+ * symbols and imports ideally. This is stored in [sorted] and is automatically computed the fist
+ * time someone accesses the property.
  */
-class ImportDependencies(tus: MutableList<TranslationUnitDeclaration>) :
-    HashMap<TranslationUnitDeclaration, MutableSet<TranslationUnitDeclaration>>() {
+class ImportDependencies<T : Node>(modules: MutableList<T>) : HashMap<T, MutableSet<T>>() {
 
     init {
-        // Populate the map with all translation units so that we have an entry in our list
+        // Populate the map with all modules so that we have an entry in our list
         // for all
-        this += tus.map { Pair(it, mutableSetOf()) }
+        this += modules.map { Pair(it, mutableSetOf()) }
     }
 
     /**
-     * The list of [TranslationUnitDeclaration] nodes, sorted by their position in the dependency
-     * graph. TUs without dependencies are first in the list, following by TUs that import TUs
-     * without dependencies, and so on.
+     * The list of [T] module nodes, sorted by their position in the dependency graph. Nodes without
+     * dependencies are first in the list, following by nodes that import nodes without
+     * dependencies, and so on.
      */
-    val sortedTranslationUnits: List<TranslationUnitDeclaration> by lazy {
-        WorkList(this).resolveDependencies()
-    }
+    val sorted: List<T> by lazy { WorkList(this).resolveDependencies() }
 
     /** Adds a dependency from [importer] to [imported]. */
-    fun add(importer: TranslationUnitDeclaration, imported: TranslationUnitDeclaration): Boolean {
-        var list = this.computeIfAbsent(importer) { mutableSetOf<TranslationUnitDeclaration>() }
-        return list.add(imported)
+    fun add(importer: T, imported: T): Boolean {
+        var list = this.computeIfAbsent(importer) { mutableSetOf<T>() }
+        var added = list.add(imported)
+
+        return added
     }
 
     /**
      * A work-list, which contains a local copy of our dependency map, so that we can remove items
      * from it while determining the order.
      */
-    class WorkList(start: ImportDependencies) :
-        HashMap<TranslationUnitDeclaration, MutableSet<TranslationUnitDeclaration>>() {
+    class WorkList<T : Node>(start: ImportDependencies<T>) : HashMap<T, MutableSet<T>>() {
```

Comment:
Would it be beneficial for the performance if we would use `IdentitySet` and `IdentityHashMap`?

Reply:
That could work, I will have a look

Reply:
Seems that performance gain is negligible, 9.3s instead of 9.8 when parsing the stdlib typings. It still makes sense to be consistent to the other passes, but I am still not sure where the big performance loss comes from.

Reply:
Ok it came from the usage of the `namespaces` extension, which we used to collect sub-namespaces. This collected all children nodes in a a namespace and then filtered them. Instead we just looked at child declarations now. 1s instead of 10s in our stdlib typings test case now 🚀